runtime/evaluate.cc

```diff
@@ -372,21 +372,21 @@ class evaluator : public node_visitor {
     copy_stmt_impl(context, *src, *dst, *op);
   }
 
+  template <typename OP>
+  void copy_dim_from(OP* op, std::size_t i, slinky::dim& dim) {
```

Comment:
I think the helper function should be: `slinky::dim eval_dim(const dim_expr& e)`, and the indexing logic should be in the caller.